src/misc/JsonConfig.cpp

```diff
@@ -0,0 +1,187 @@
+
+namespace fs = std::filesystem;
+
+JsonConfig::JsonConfig(std::string path) : mPath(path) {
+    // check if path exists and that it's a file
+    if (!fs::exists(path) || !fs::is_regular_file(path)) {
+        mIsNewConfig = true;
+        mJson = json::object();
+        return;
+    }
+
+    try {
+        std::ifstream ifs(path);
+        mJson = json::parse(ifs);
+    } catch (json::parse_error& e) {
+        SPDLOG_ERROR("Failed to parse JSON file: {}", e.what());
+
+        // If failure to parse, we will create a new file
+        mIsNewConfig = true;
+        mJson = json::object();
+    }
+}
+
+void JsonConfig::SetArbitraryType(std::string key, json::value_type value) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    if (keyParts.size() > 1) {
+        // nested key
+        json* currentJson = &mJson;
+        for (size_t i = 0; i < keyParts.size() - 1; i++) {
+            if (!currentJson->contains(keyParts[i])) {
+                // create new object
+                (*currentJson)[keyParts[i]] = json::object();
+            }
+            currentJson = &(*currentJson)[keyParts[i]];
+        }
+        (*currentJson)[keyParts[keyParts.size() - 1]] = value;
+    } else {
+        // not nested
+        mJson[key] = value;
+    }
+}
+
+json::value_type JsonConfig::GetArbitraryType(std::string key) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    // find the deepest nested key if it exists
+    json* currentJson = &mJson;
+    for (size_t i = 0; i < keyParts.size() - 1; i++) {
+        if (!currentJson->contains(keyParts[i])) {
+            return nullptr;
+        }
+        currentJson = &(*currentJson)[keyParts[i]];
+    }
+
+    // extract the value if possible
+    if (currentJson->contains(keyParts[keyParts.size() - 1])) {
+        return (*currentJson)[keyParts[keyParts.size() - 1]];
+    } else {
+        return nullptr;
+    }
+}
+
+// MARK: - Public API
+
+bool JsonConfig::IsNewConfig() {
+    return mIsNewConfig;
+}
+
+void JsonConfig::DeleteEntry(std::string key) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    // clear nested key, if after deletion parent is empty, delete it too
+    if (keyParts.size() > 1) {
+        json* currentJson = &mJson;
+        for (size_t i = 0; i < keyParts.size() - 1; i++) {
+            if (!currentJson->contains(keyParts[i])) {
+                return;
+            }
+            currentJson = &(*currentJson)[keyParts[i]];
+        }
+        (*currentJson).erase(keyParts[keyParts.size() - 1]);
+
+        // delete empty parents
+        for (size_t i = keyParts.size() - 1; i > 0; i--) {
+            if ((*currentJson).empty()) {
+                currentJson = &mJson;
+                for (size_t j = 0; j < i - 1; j++) {
+                    currentJson = &(*currentJson)[keyParts[j]];
+                }
+                (*currentJson).erase(keyParts[i - 1]);
+            }
+        }
+    } else {
+        // not nested
+        mJson.erase(key);
+    }
```

Comment:
i'd invert this to avoid having everything deep in the `if`, so just

```cpp
if (keyParts.size() == 1) {
    // not nested
    mJson.erase(key);
    return;
}

[everything from the if block]
```